BJS2-84097

mrnght opened the pull request:
Реализовал асинхронное удаление прошедших событий

mrnght2 committed:
BJS2-84097
Реализовал асинхронное удаление прошедших событий

RinatBeybutov commented on src/main/java/school/faang/user_service/scheduler/Scheduler.java:
```diff
@@ -0,0 +1,17 @@
+    private final EventService service;
+
+    @Scheduled(cron = "${event.delete-scheduled.cron}")
+    private void clearEvents() {
```
ты тестировал это?

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventServiceImpl.java:
```diff
@@ -92,6 +99,37 @@ public void delete(long eventId) {
+        List<Long> pastEventsIds = events.stream()
+                .filter(event -> event.getStatus().equals(EventStatus.COMPLETED))
+                .map(event -> event.getId())
+                .toList();
```
здесь неоптимальное решение, ты сначала ищешь все записи, а потом фильтруешь в памяти. Лучше написать запрос, который достает все ID эвентов, которые в статусе Completed. И выглядеть будет лаконично и лучше работать

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventServiceImpl.java:
```diff
@@ -92,6 +99,37 @@ public void delete(long eventId) {
+                .toList();
+
+        int threadCount = (int) Math.ceil((double) pastEventsIds.size() / eventBucketSize);
+        ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
```
создание нового пула потоков для каждой операции не ок, лучше создать один раз

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventServiceImpl.java:
```diff
@@ -92,6 +99,37 @@ public void delete(long eventId) {
+        for (int i = 0; i < pastEventsIds.size(); i += eventBucketSize) {
+            List<Long> sublist = pastEventsIds.subList(i,
+                    Math.min(i + eventBucketSize, pastEventsIds.size()));
+            executorService.submit(() -> eventRepository.deleteByIds(sublist));
```
если во время удаления возникнет ошибка, то у тебя получится состояние когда часть удалилась, а часть - нет

mrnght replied:
Я не понимаю, какая ошибка тут может возникнуть и как решить указанную тобой проблему. Сделать логику доп. попыток?

RinatBeybutov replied:
ошибка связанная с базой, например таймаут по запросу на удаление и последующий роллбэк транзакции, чтобы все ок было по архитектуре, можно:
1. Создать в сервисе метод получения всех нужных id, 
2. Потом в классе Scheduler разделить на группы для каждой из них вызывать метод сервиса на удаление, тогда и транзакция будет работать, и при падении одной группы, остальные не упадут по цепочке. Кроме того, неудаленные записи могут удалиться в следующую очистку

mrnght2 committed:
BJS2-84097
Перенес метод в репозиторий

mrnght2 committed:
BJS2-84097
пофиксил любимый чекстайл

RinatBeybutov commented on src/main/java/school/faang/user_service/scheduler/Scheduler.java:
```diff
@@ -11,7 +11,8 @@ public class Scheduler {
     @Scheduled(cron = "${event.delete-scheduled.cron}")
-    private void clearEvents() {
+    public void clearEvents() {
+        System.out.println("вызов метода");
```
нужен какой-то значимый лог с использованием логгера

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventServiceImpl.java:
```diff
@@ -92,6 +101,34 @@ public void delete(long eventId) {
+            executorService.submit(() -> eventRepository.deleteByIds(sublist));
+        }
+
+        executorService.shutdown();
```
а вот прерывать его не стоит, потому как воспользоваться им второй раз не получится

mrnght2 committed:
BJS2-84097
Убрал закрытия пулла потоков и вынес логику разбития событий на бакеты в класс Scheduler

RinatBeybutov commented:
поправь стиль кода и можно мержить

mrnght2 committed:
BJS2-84097
удалил неиспользуемые импорты
